src/components/display/CsvDisplay.tsx

```diff
@@ -0,0 +1,89 @@
+import { useEffect, useState } from "react";
+import Papa from "papaparse";
+import { Alert } from "antd";
+
+import SpreadsheetTable, { SPREADSHEET_ROW_KEY_PROP, SpreadsheetTableProps } from "./SpreadsheetTable";
+
+const DEFAULT_COLUMN = { key: "col" };
+
+type CsvDisplayProps = {
+  contents?: Blob | null;
+  loading?: boolean;
+};
+
+type CsvRecord = Record<string, string>;
+type CsvData = CsvRecord[];
+
+type CsvParseResult = [SpreadsheetTableProps<CsvRecord>["columns"], CsvData];
+
+const CsvDisplay = ({ contents, loading }: CsvDisplayProps) => {
+  const [parsedData, setParsedData] = useState<CsvData>([]);
+  const [parseError, setParseError] = useState("");
+  const [isParsing, setIsParsing] = useState(true); // Start in parsing state
+  const [columns, setColumns] = useState<SpreadsheetTableProps<CsvRecord>["columns"]>([DEFAULT_COLUMN]);
+
+  useEffect(() => {
+    if (contents === undefined || contents === null) return;
+
+    setIsParsing(true);
+
+    contents
+      .text()
+      .then(
+        (csvText) =>
+          new Promise<CsvParseResult>((resolve, reject) => {
+            const rows: CsvData = [];
+
+            let innerParseError: string | null = null;
+
+            // noinspection JSUnusedGlobalSymbols
+            Papa.parse<string[]>(csvText, {
+              worker: true,
+              step: (res) => {
+                if (res.errors?.length) {
+                  innerParseError = res.errors[0].message;
+                }
```

Comment:
(very) minor: in case of an error on an early row, does it keep parsing rows until the end of the document?
If so, it would be better to have an early reject on the first error encountered, since we only show an alert in that case.

Reply:
since the parsing isn't done as a promise, early reject may mean it keeps parsing after promise rejection, which might get messy. I'll look into the best thing to do here.

Reply:
for now I just aggregate multiple errors as it goes along